Source/Core/Core/PowerPC/Jit64/Jit.cpp

```diff
@@ -719,9 +720,11 @@ const u8* Jit64::DoJit(u32 em_address, PPCAnalyst::CodeBuffer *code_buf, JitBloc
 		bool gatherPipeIntCheck = js.fifoWriteAddresses.find(ops[i].address) != js.fifoWriteAddresses.end();
 
 		// Gather pipe writes using an immediate address are explicitly tracked.
-		if (jo.optimizeGatherPipe && js.fifoBytesThisBlock >= 32)
+		if (jo.optimizeGatherPipe && (js.fifoBytesThisBlock >= 32 || js.mustCheckFifo))
 		{
-			js.fifoBytesThisBlock -= 32;
+			if (js.fifoBytesThisBlock >= 32)
+				js.fifoBytesThisBlock -= 32;
```

Comment:
After reading FastCheckGatherPipe, I think you're right. Seting it to zero is better in the long term. But yeah, splitting this in another PR is likely a wise idea.